Feature/highlights

dida0423 opened the pull request:
<img width="1430" alt="image" src="https://github.com/user-attachments/assets/b3ce2c8f-3285-4d3d-852d-12f871ded63d" />

- [POST] /highlight/new/{story_id} - 새로운 Highlight (Parameters: Highlight name, Cover image) 생성 및 Story 저장
  - Cover image 없을 시 스토리 이미지-> 커버 이미지 (스토리 여러개 있는 상태에서 커버 Story 삭제할 경우 exception handling 필요할 듯) (TODO)
  - Story 없이 Highlight 생성 불가
- [PATCH] /highlight/add/{highlight_id}/{story_id} - 이미 존재하는 highlight에 story ID로 story 추가
- [GET] /highlights/{user_id} - 특정 유저의 모든 Highlight 반환
- [GET] /highlight/{highlight_id} - 특정 Highlight 반환
- [DELETE] /highlight/{highlight_id} - 특정 Highlight 삭제
- [DELETE] /highlightstory/{highlight_id}/{story_id} 특정 Highlight에서 Story 삭제
  - 모든 Story 삭제 시 Highlight 삭제 처리 필요 (TODO)

dida0423 committed:
feat: highlight model & create highlight

dida0423 committed:
feat: create new highlight

dida0423 committed:
feat: add story to highlight

dida0423 committed:
feat: delete highlight

dida0423 committed:
feat: unsave story

dida0423 committed:
fix: delete

dida0423 committed:
fix: story highlight relation on delete

cloNoey committed:
Merge branch 'main' into feature/highlights

cloNoey commented on instaclone/app/story/views.py:
```diff
@@ -67,6 +67,75 @@ async def delete_story(
+    if type(highlight_create_request.cover_image) == UploadFile:
+        cover_image = await medium_service.create_medium(highlight_create_request.cover_image)
+    else:
+        story = await story_service.get_story(story_id=story_id)
```
```suggestion
        story = await story_service.get_story(story_id=story_id, current_user=user)
```

cloNoey reviewed (approved):
수고하셨습니다! 
1. story_service.get_story(~) 에 current_user 변수 추가하도록 수정해주시면 될 것 같습니다.
2. Add To Highlight에서 이미 하이라이트에 등록한 스토리의 경우, 해당 하이라이트에 다시 등록되지 못하도록 제한하는 예외처리도 추가해주시면 좋을 것 같아요! (프론트에서 해결할 수도 있을 것 같지만, 복잡하지 않다면 예외처리 해주시면 더 좋을 듯 합니다 :) )

mjy926 commented on instaclone/app/story/store.py:
```diff
@@ -92,10 +95,161 @@ async def delete_story(
+
+        if not highlight:
+            raise HighlightDNEError()
+        if story.user_id != user.user_id or highlight.user_id != user.user_id:
+            raise StoryPermissionError()
```
스토리를 삭제할 권한이 없다고 나와서 커스텀 에러 새로 정의해도 좋을 것 같아요. StoryPermissionError로 에러 처리되어 있는 부분에서 모두 같은 에러 메세지가 나와요.

mjy926 reviewed (approved):
수고하셨습니다!

dida0423 committed:
fixes

- Deleting story removes story from highlight
- Removing story with cover image from highlight selects new image from highlight
- Removing last story from highlight deletes highlight
- Highlight of the same name as the user's existing highlights cannot be created
- Same story cannot be added to the same story twice

cloNoey committed:
Merge branch 'main' into feature/highlights

dida0423 commented:
@cloNoey 감사합니다!

혹시 인스타그램에선 한 스토리를 여러 하이라이트로 올릴 수 없나요?
